wincent/liferay-portal
# LPS-108803 Disabling modules preloading does not work

Hey @wincent ,

To test the code live:
1. Delete `generateUI = false` from [this line](https://github.com/brianchandotcom/liferay-portal/blob/00360fa575c37208ef7fb3c9e48203e38d806ab1/modules/apps/frontend-js/frontend-js-aui-web/src/main/java/com/liferay/frontend/js/aui/web/internal/configuration/AUIConfiguration.java#L24) and redeploy
2. Control Panel > Configuration > System Settings >  Third Party > AUI > Preload Widely Used AUI Modules
3. With preload option checked, [files](https://github.com/brianchandotcom/liferay-portal/blob/00360fa575c37208ef7fb3c9e48203e38d806ab1/modules/apps/frontend-js/frontend-js-aui-web/src/main/java/com/liferay/frontend/js/aui/web/internal/servlet/AUITopHeadResources.java#L89) under `_FILE_NAMES_AUI_PRELOAD` will appear in document.head. If it is unchecked, they will not, and there should be no errors.

For start, this PR extends AUI Core files to include `aui_sandbox.js`, as it adds AUI global variable. This is a simple and clear change.

Much more difficult is the issue that we have eager code executions of functions defined in eagerly executed AUI-dependent code. These are offenders:
1. `Liferay.provide`
1. Custom event definitions, `Liferay.on`, `Liferay.publish` etc.
1. `Liferay.BrowserSelectors.run`
1. Utilities in `util.js`

This PR fixes that issue in the simplest possible way: by extending AUI Core definition to include everything those eager executions require: files with definitions (for example, `dependency.js` that defines `Liferay.provide`) and requirements of those definitions (for example, `oop.js` used in `dependency.js`). This has the effect that code like [custom events augment](https://github.com/brianchandotcom/liferay-portal/blob/00360fa575c37208ef7fb3c9e48203e38d806ab1/modules/apps/frontend-js/frontend-js-web/src/main/resources/META-INF/resources/liferay/events.js#L20) is executed immediately, instead of lazily loading dependencies while custom events definitions are breaking.

Notes:
1. The custom events augment is specially offending, as it is not even in a dedicated file, but in a self-executing code in the wrong module (`frontend-js-web`). Fortunately, as it is in `frontend-js-web`, it gets executed before other portal eager usages, so `events.js` does not need to be a part of AUI core. For clarity, this line should probably have its own file in `frontend-js-aui-web`, but the proper solution is to rewrite it with own custom events handling.
1. We could replace eager definitions with proper AUI modules, and wrap all usages in `A.use()` of that module. This is probably too much work, and a bit invasive.
1. If the solution in this PR is ok, we may want to rephrase control panel descriptions, as AUI Core is not just global `AUI`.

Please review the code.

Thank you!

/cc @izaera @jbalsas
